ports/raspberrypi/common-hal/rp2pio/Pio.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Scott Shawcroft for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "common-hal/rp2pio/Pio.h"
+#include "py/gc.h"
+#include "src/rp2_common/hardware_irq/include/hardware/irq.h"
+
+PIO all_pios[NUM_PIOS] = { pio0, pio1 };
+
+typedef struct {
+    rp2pio_pio_irq_handler_t handler;
+    void *context;
+} rp2pio_pio_irq_t;
+
+STATIC rp2pio_pio_irq_t *_irq_table[NUM_PIOS] = { NULL, NULL };
+
+STATIC rp2pio_pio_irq_t *_get_irq_entry(PIO pio, enum pio_interrupt_source source) {
+    uint index = pio_get_index(pio);
+    return &_irq_table[index][source];
+}
+
+STATIC void _irq_handler(PIO pio) {
+    for (int i = 0; i < NUM_PIO_INTERRUPT_SOURCES; i++) {
+        if (pio->ints0 & (1 << i)) {
+            pio_set_irq0_source_enabled(pio, i, false);
+            rp2pio_pio_irq_t *entry = _get_irq_entry(pio, i);
+            entry->handler(pio, i, entry->context);
+        }
+    }
+}
+
+STATIC void _irq_handler_pio0(void) {
+    _irq_handler(pio0);
+}
+
+STATIC void _irq_handler_pio1(void) {
+    _irq_handler(pio1);
+}
+
+STATIC void _cinit_pio(uint pio_index, uint irq, irq_handler_t irq_handler) {
+    assert(pio_index < NUM_PIOS);
+    PIO pio = all_pios[pio_index];
+    rp2pio_pio_irq_t **irq_table = &_irq_table[pio_index];
+    if (*irq_table != NULL) {
+        return;
+    }
+
+    pio_clear_instruction_memory(pio);
+    *irq_table = m_new_ll(rp2pio_pio_irq_t, NUM_PIO_INTERRUPT_SOURCES);
```

Comment:
this probably didn't do what you intended.  the 2nd argument states whether the object should be allocated in the long-lived portion of the heap. it's not giving you NUM_PIO_INTERRUPT_SOURCES objects of type rp2pio_pio_irq_t.

Reply:
Are you sure?
```#define m_new_ll(type, num) ((type *)(m_malloc(sizeof(type) * (num), true)))```

Reply:
yets, you're right after all. I must have looked at a different macro definition.